File: python/may/trie.py
class TrieNode:
    def __init__(self):
        self.is_word = False
        self.children = {}


class Trie:

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str):
        ws = self.root
        for i in word:
            if i not in ws.children:
                ws.children[i] = TrieNode()
            ws = ws.children[i]
        ws.is_word = True

    # @param {string} word
    # @return {boolean}
    # Returns if the word is in the trie.
    def search(self, word):
        node = self.root
        for i in word:
            if i not in node.children:
                return False
            node = node.children[i]
        return node.word

    # @param {string} prefix
    # @return {boolean}
    # Returns if there is any word in the trie
    # that starts with the given prefix.
    def starts_with(self, prefix):
        node = self.root
        for i in prefix:
            if i not in node.children:
                return False
            node = node.children[i]
        return True

File: go/other/trie/trie.go
package trie

type Trie struct {
	Children [26]*Trie
	IsWord   bool
}

func (trie *Trie) Insert(word string) {
	node := trie
	for i := 0; i < len(word); i++ {
		c := word[i]
		if node.Children[c-'a'] == nil {
			node.Children[c-'a'] = new(Trie)
		}
		node = node.Children[c-'a']
	}
	node.IsWord = true
}

func (trie *Trie) Search(word string) bool {
	node := trie
	for i := 0; i < len(word); i++ {
		c := word[i]
		if node.Children[c-'a'] == nil {
			return false
		}
		node = node.Children[c-'a']
	}
	return node.IsWord
}

func (trie *Trie) StartsWith(word string) bool {
	node := trie
	for i := 0; i < len(word); i++ {
		c := word[i]
		if node.Children[c-'a'] == nil {
			return false
		}
		node = node.Children[c-'a']
	}
	return true
}
